src/schemas.jl

```diff
@@ -449,11 +493,31 @@ function _generate_validation_definitions(schema_version::SchemaVersion)
             isnothing(result) || return $fname => result
```

Comment:
this seems to be looping over all violations so I guess the issue here is that when the  `return` is evaluated it exits the loop early and so doesn't show all results?

Reply:
played around a MWE locally and this seems to be the case

Reply:
yeah, exactly---which I suspect could make for significantly faster validation for cases where complete, user-readable errors are less important (if the various schemas are huge and/or complex)? I think that's perhaps a reason not do do:

> I wonder if we could now delete the code for find_violation and simply overload it with find_violation(x...) = first(find_violations(x...)) ?
> And maybe even @deprecate it with the view to removing it altogether in the next breaking release!

For example, `complies_with` uses this `find_violation` (singular!) approach. I _would_ like to rename this "`find_first_violation`" or something, but am not sure it's worth the breaking change? Or maybe it's worth following your proposed deprecation path but for the rename...

> Having two functions do basically the same thing but only differ semantically has potential for headaches down the line...in particular if they started to diverge functionally or one introduced a bug, etc

Yeah, this is a good point. It would be nicer if the macros that generate the two shared code---and tbh, the reason that they don't is that I'm so new to this macro-on-macro style that it was tricky enough for me to figure out how do do them separately! I'm not _too_ worried that they diverge yet, given that we have tests for both of them, and they're close enough in code location that it seems likely that someone adjusting one would see the other as well, but this is a great point. Let me do a quick timebox to attempt sharing code, now that I understand this code better---if I can't figure something out quickly, I think we could leave it for the future (or someone else could propose a fix here! :) ).

Reply:
Right okay, refactored! Not sure it's easier to read, but at least they share a single codepath now...?

Does anyone have strong opinions re: the rename/deprecation?

Reply:
Thanks for refactoring - I think sharing code is definitely a positive step!

> I suspect could make for significantly faster validation for cases where complete, user-readable errors are less important (if the various schemas are huge and/or complex)?

Hmm...how large/complex do these schemas get? And would the cost of exhaustively running `_check_for_expected_field` really be significant?

I won't let this bike-shedding block the PR but I would be surprised to learn if it really was a computational bottleneck! 
If not, it would be trivial to switch `complies_with` to check `isempty(find_violations(...))`